### PR TITLE
Gmail: Search large numbers of mailboxes

### PR DESCRIPTION
**Pull request by israelpoli**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4352

## Description
Added support large search for companies with over 2500 accounts.
In searching over 500 to 2500 accounts the use can be done by a command,
Over 2500 accounts the search will be done by playbook.

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 1]
```diff
@@ -745,6 +757,100 @@ def get_millis_from_date(date, arg_name):
...
+
+    return accounts
+
+
+def scheduled_commands_for_more_users(accounts: list, next_page_token: str) -> List[CommandResults]:
+
+    accounts_batches = cutting_for_batches(accounts)
+
+    command_results: List[CommandResults] = []
+    args = {k: v for k, v in demisto.args().items()}
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:780]: why not just `demisto.args()`?

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 2]
```diff
@@ -745,6 +757,100 @@ def get_millis_from_date(date, arg_name):
...
+        accounts[0].extend(list_accounts[batch_size * BATCH_DIVIDER:])
+
+    return accounts
+
+
+def scheduled_commands_for_more_users(accounts: list, next_page_token: str) -> List[CommandResults]:
+
+    accounts_batches = cutting_for_batches(accounts)
+
+    command_results: List[CommandResults] = []
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:779]:
```suggestion
    command_results: list[CommandResults] = []
```
python3.10 :)

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 3]
```diff
@@ -745,6 +757,100 @@ def get_millis_from_date(date, arg_name):
...
+    args = {k: v for k, v in demisto.args().items()}
+    for batch in accounts_batches:
+
+        args.update({'list_accounts': batch})
+        command_results.append(
+            CommandResults(
+                readable_output='Please wait... serching...',
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:786]:
```suggestion
                readable_output='Serching mailboxes, please wait...',
```

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 4]
```diff
@@ -745,6 +757,100 @@ def get_millis_from_date(date, arg_name):
...
+                scheduled_command=ScheduledCommand(
+                    command='gmail-search-all-mailboxes',
+                    next_run_in_seconds=10,
+                    args={k: v for k, v in args.items()},
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:790]: why not just pass `args`?

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 5]
```diff
@@ -745,6 +757,100 @@ def get_millis_from_date(date, arg_name):
...
+            break
+
+    return list_accounts, users_next_page_token
+
+
+def information_search_process(length_accounts: int, searching_accounts: str) -> CommandResults:
+
+    if not searching_accounts:
+        readable_output = f'Searching accounts, ' \
+                          f'from 1 to {length_accounts} from all the accounts'
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:840]:
```suggestion
        readable_output = f'Searching the first {length_accounts} accounts'
```

File: Packs/Gmail/Integrations/Gmail/Gmail.yml [hunk 6]
```diff
@@ -623,16 +623,27 @@ script:
       secret: false
     - auto: PREDEFINED
       default: false
-      description: Whether to return only mailboxes which contain the email. Default is "False".
+      description: Whether to return only mailboxes which contain the email. Default is "True".
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.yml:626]: isn't it `false`, in line 625?

File: Packs/Gmail/Integrations/Gmail/Gmail.yml [hunk 7]
```diff
@@ -623,16 +623,27 @@ script:
...
       name: show-only-mailboxes
       predefined:
       - "false"
       - "true"
       required: false
       secret: false
+    - default: false
+      description: This is a list of accounts where the search will take place.
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.yml:634]:
```suggestion
      description: Accounts to search in.
```

File: Packs/Gmail/Integrations/Gmail/Gmail.yml [hunk 8]
```diff
@@ -623,16 +623,27 @@ script:
...
+      required: false
+      secret: false
+    - default: false
+      description: Used to track a search's progress.
+      name: searching_accounts
+      required: false
+      secret: false
     deprecated: false
-    description: Searches the Gmail records for all Google users.
+    description: Searches the Gmail records for all Google users. (When there is more than 2500 accounts please use with the Search all mailboxes - Gmail with polling playbook)
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.yml:644]:
```suggestion
    description: Searches the Gmail records for all Google users. (For more than 2500 accounts, please use with the `Search all mailboxes - Gmail with polling` playbook)
```

File: Packs/Gmail/Integrations/Gmail/Gmail.yml [hunk 9]
```diff
@@ -671,9 +682,18 @@ script:
     - contextPath: Gmail.Headers
       description: All headers of specific mail (list).
       type: unknown
-    - contextPath: Gmail.Mailbox
-      description: The Gmail Mailbox.
-      type: string
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.yml:676]: is this a breaking change?

File: Packs/Gmail/Integrations/Gmail/Gmail.yml [hunk 10]
```diff
@@ -671,9 +682,18 @@ script:
...
+    - contextPath: Gmail.Mailboxes
+      description: The Gmail mailboxes in which emails matching the query were found.
+      type: String
+    - contextPath: Gmail.UnsearchedAcounts
+      description: The unsearched acounts with the error.
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.yml:689]: Please consult with @ShirleyDenkberg on this term

File: Packs/Gmail/ReleaseNotes/1_2_7.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
+
+#### Integrations
+##### Gmail
+- **Note**: The argument *show-only-mailboxes* will now be True by default when using the command ***gmail-search-all-mailboxes***.
+- Added support for a large search in the ***gmail-search-all-mailboxes*** command, please see to the description of the command.
```
Comment (dorschw) [inline Packs/Gmail/ReleaseNotes/1_2_7.md:5]:
```suggestion
- Added support for searching large numbers of mailboxes in ***gmail-search-all-mailboxes***, please refer to the command's description for more information.
```

File: Packs/Gmail/ReleaseNotes/1_2_7.md [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+
+#### Playbooks
+##### New: Search all mailboxes - Gmail with polling
+- Added the new playbook. This playbook searches Gmail records for all Google users, designed for large companies with over 2500 Google users. (Available from Cortex XSOAR 6.5.0).
```
Comment (dorschw) [inline Packs/Gmail/ReleaseNotes/1_2_7.md:9]: Does it search _all_ accounts accessible?

File: Packs/Gmail/pack_metadata.json [hunk 13]
```diff
@@ -2,7 +2,7 @@
     "name": "Gmail",
     "description": "Gmail API and user management (This integration replaces the Gmail functionality in the GoogleApps API and G Suite integration).",
     "support": "xsoar",
-    "currentVersion": "1.2.6",
+    "currentVersion": "1.2.7",
```
Comment (dorschw) [inline Packs/Gmail/pack_metadata.json:5]:
```suggestion
    "currentVersion": "1.3.0",
```

Please change the RN filename as well

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 14]
```diff
@@ -745,6 +757,100 @@ def get_millis_from_date(date, arg_name):
             raise ValueError(f'{arg_name} argument is not in expected format.')
 
 
+def cutting_for_batches(list_accounts: list) -> List[list]:
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:760]: Please use `batch()` from CommonServerPython instead

**Review by dorschw — COMMENTED** [review]
Impressive, please see the comments

File: Packs/Gmail/Integrations/Gmail/Gmail.yml (hunk 10 above)
Comment (julieschwartz18) [reply Packs/Gmail/Integrations/Gmail/Gmail.yml:690]: @israelpoli - can you please explain what you mean in the description "The unsearched accounts with the error." - what error? Is this supposed to be a list of Gmail accounts? Also, the contextPath should be spelled Gmail.UnsearchedAccounts (accounts with 2 c's)

File: Packs/Gmail/ReleaseNotes/1_3_0.md [hunk 15]
```diff
@@ -0,0 +1,10 @@
+
+#### Integrations
+##### Gmail
+- **Note**: The argument *show-only-mailboxes* will now be True by default when using the command ***gmail-search-all-mailboxes***.
+- Added support for searching large numbers of mailboxes in ***gmail-search-all-mailboxes***, please refer to the command's description for more information.
+#### Playbooks
+##### New: Search all mailboxes - Gmail with polling
+- Added the new playbook. This playbook searches Gmail records for all Google users, designed for large companies with over 2500 Google users. (Available from Cortex XSOAR 6.5.0).
+##### New: Search in mailboxes Gmail (Loop) with polling
+- Added the new playbook. This playbook is used as a subplaybook for the Playbook Search-all-mailboxes - Gmail, it should not be used. (Available from Cortex XSOAR 6.5.0).
```
Comment (julieschwartz18) [inline Packs/Gmail/ReleaseNotes/1_3_0.md:10]: @israelpoli - what do you mean by "it should not be used" - what should not be used?

Comment (israelpoli) [reply Packs/Gmail/ReleaseNotes/1_3_0.md:10]: @julieschwartz18 

It's a subplaybook that when you run it alone it crashes,
Run this subplaybook only through the parent playbook.

File: Packs/Gmail/Integrations/Gmail/Gmail.yml [hunk 16]
```diff
@@ -671,9 +683,18 @@ script:
...
+      type: String
+    - contextPath: PageToken.NextPageToken
+      description: The use of this output is in the Search all mailboxes - Gmail with polling Playbook, so NextPageToken is used to call back the command to fetch the list of the next accounts in which the search will be made.
```
Comment (julieschwartz18) [inline Packs/Gmail/Integrations/Gmail/Gmail.yml:693]: Updated this to "Fetches the next list of accounts that will be searched." - is this ok?

File: Packs/Gmail/Integrations/Gmail/Gmail.yml [hunk 17]
```diff
@@ -671,9 +683,18 @@ script:
...
+      type: String
+    - contextPath: SearchProcess
+      description: The use of this output is in the Search all mailboxes - Gmail with polling Playbook, it allows the search process to be followed.
```
Comment (julieschwartz18) [inline Packs/Gmail/Integrations/Gmail/Gmail.yml:696]: Updated this to "Enables the search process to be followed." - is this ok?

File: Packs/Gmail/Integrations/Gmail/Gmail.yml (hunk 16 above)
Comment (israelpoli) [reply Packs/Gmail/Integrations/Gmail/Gmail.yml:693]: yes

File: Packs/Gmail/Integrations/Gmail/Gmail.yml (hunk 17 above)
Comment (israelpoli) [reply Packs/Gmail/Integrations/Gmail/Gmail.yml:696]: yes

**Comment by julieschwartz18** [comment]
Doc review complete for the Gmail.yml file. @israelpoli - please confirm the following.
- line 690: description for Gmail.UnsearchedAcounts is "The unsearched accounts with the error." - this is unclear, can you explain?
- line 693: updated the description for PageToken.NextPageToken - please confirm
- line 696: updated the description for SearchProcess - please confirm
- line 768: the description for query is "Query string search. Should be of the form "". Complete documentation is at https://developers.google.com/admin-sdk/directory/v1/guides/search-users" - the form is empty quotes, is this correct?
- line 833: Account.Email.Adderss context path is misspelled (should be Account.Email.Address)

**Comment by julieschwartz18** [comment]
Doc review complete for the [playbook-Search-all-mailboxes_with_polling.yml](https://github.com/demisto/content/pull/23175/commits/de6ab12009c84e1b70481337391e76a8ce24dac0) file. @israelpoli - please confirm the edits. Specifically, please confirm the updated playbook description on line 115.

**Comment by julieschwartz18** [comment]
Doc review complete for the [playbook-Search-in-mailboxes_Loop_with_polling.yml](https://github.com/demisto/content/pull/23175/commits/9f87e863a22ed0d5d0dbd31982352f3835f5e634) file. @israelpoli - please confirm the description updates.

**Comment by julieschwartz18** [comment]
@israelpoli - there are updates to descriptions throughout the yaml files, please regenerate the READMEs. Thanks!

**Comment by julieschwartz18** [comment]
Doc review complete for the Gmail README file. @israelpoli - please review the updates and confirm.  Also:
- line 81: the description for gapps is "The immutable Google Apps ID -  Only the Cxxxxxxxx, section is needed." - can you explain this?

**Comment by julieschwartz18** [comment]
Doc review complete for this file. @israelpoli - please confirm:
- line 60: the description for the SearchProcess output is "Needs rewrite." - this is unclear, can you explain?

**Comment by israelpoli** [comment]
> Doc review complete for this file. @israelpoli - please confirm:
> 
> * line 60: the description for the SearchProcess output is "Needs rewrite." - this is unclear, can you explain?

Fixed

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 18]
```diff
@@ -35,7 +36,7 @@
 
 ''' GLOBAL VARS '''
 
-ADMIN_EMAIL = None
+ADMIN_EMAIL = ''
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:39]:
```suggestion
ADMIN_EMAIL = '' # set from params later on
```

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 19]
```diff
@@ -404,21 +409,29 @@ def create_incident_labels(parsed_msg, headers):
     return labels
 
 
-def mailboxes_to_entry(mailboxes):
+def mailboxes_to_entry(mailboxes: list) -> list[CommandResults]:
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:412]:
```suggestion
def mailboxes_to_entry(mailboxes: list[dict]) -> list[CommandResults]:
```
right?

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 20]
```diff
@@ -745,12 +758,109 @@ def get_millis_from_date(date, arg_name):
...
 
+def cutting_for_batches(list_accounts: list) -> List[list]:
+
+    accounts: list = []
+    rest_accounts: list = []
+
+    batch_size = int(len(list_accounts) / BATCH_DIVIDER)
+    if rest := len(list_accounts) % BATCH_DIVIDER:
+        rest_accounts = list_accounts[-rest:]
+        list_accounts = list_accounts[:-rest]
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:769]: isn't this handled in `batch` (from CommonServerPython)?

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 21]
```diff
@@ -745,12 +758,109 @@ def get_millis_from_date(date, arg_name):
...
+
+    accounts.extend(batch(list_accounts, batch_size))
+
+    if rest_accounts:
+        accounts[0].extend(rest_accounts)
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:774]: why [0]?

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 22]
```diff
@@ -745,12 +758,109 @@ def get_millis_from_date(date, arg_name):
...
+
+    return command_results
+
+
+def get_mailboxes(max_results: int, users_next_page_token: str = None):
+    '''
+    Used to fetch the list of accounts for the search-all-mailboxes command
+    '''
+    list_accounts: list[str] = []
+    counter = 0
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:816]: what are we counting? Let's name it `<...>_count` according to the thing being counted

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 23]
```diff
@@ -1146,55 +1256,69 @@ def get_user_tokens(user_id):
...
-def search_all_mailboxes(receive_only_accounts):
-    users_next_page_token = None
-    service = get_service('admin', 'directory_v1')
-    while True:
-        command_args = {
-            'maxResults': 100,
-            'domain': ADMIN_EMAIL.split('@')[1],  # type: ignore
-            'pageToken': users_next_page_token
-        }
+def search_in_mailboxes(list_accounts: list[str], receive_only_accounts: bool) -> None:
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:1259]: 1. Please add docstrings
2. If `list_accounts` is a list of accounts, a better name would be `accounts`.
3. `receive_only_accounts` is unclear

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 24]
```diff
@@ -1146,55 +1256,69 @@ def get_user_tokens(user_id):
...
-        users_count = len(result['users'])
-        users_next_page_token = result.get('nextPageToken')
+    futures: list = []
+    entries: list = []
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        for user in list_accounts:
+            futures.append(executor.submit(search_command, mailbox=user,
+                                           receive_only_accounts=receive_only_accounts))
+        for account in concurrent.futures.as_completed(futures):
+            if found := account.result():
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:1268]: what if it's not found?

File: Packs/Gmail/Integrations/Gmail/Gmail.py [hunk 25]
```diff
@@ -1146,55 +1256,69 @@ def get_user_tokens(user_id):
...
+                entries.append(found)
 
-        with concurrent.futures.ThreadPoolExecutor() as executor:
-            accounts_counter = 0
-            futures = []
+        if entries:
+            if receive_only_accounts:
+                entries = [mailboxes_to_entry(entries)]
+            return_results(entries)
             entries = []
```
Comment (dorschw) [inline Packs/Gmail/Integrations/Gmail/Gmail.py:1275]: is this left here on purpose?

File: Packs/Gmail/Integrations/Gmail/Gmail.py (hunk 24 above)
Comment (israelpoli) [reply Packs/Gmail/Integrations/Gmail/Gmail.py:1268]: It returns None so I don't collect the results in that case

Comment (dorschw) [reply Packs/Gmail/Integrations/Gmail/Gmail.py:1268]: please mention it in the docstring of this method

**Pull request merged**